### PR TITLE
Mark ZST as FFI-safe if all its fields are PhantomData

### LINKED ISSUE #106629: `improper_ctypes` lint fires on `#[repr(transparent)]` wrappers around `PhantomData` (in struct fields).
If I have a `#[repr(transparent)]` wrapper around a `PhantomData`, it's considered an improper C type when used in a struct field, even though the wrapped `PhantomData` would be allowed. Specifically ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c82e16eb77a88e4613e5812d1b1123e3))
```rs
#[repr(transparent)]
#[derive(Copy, Clone)]
struct MyPhantom(core::marker::PhantomData<u8>);

#[repr(C)]
#[derive(Copy, Clone)]
pub struct Bar {
    pub x: i32,
    _marker: MyPhantom,
}

extern "C" {
    pub fn foo(bar: *mut Bar);
}
```

produces the error
```
   Compiling playground v0.0.1 (/playground)
warning: `extern` block uses type `MyPhantom`, which is not FFI-safe
  --> src/lib.rs:13:21
   |
13 |     pub fn foo(bar: *mut Bar);
   |                     ^^^^^^^^ not FFI-safe
   |
   = note: this struct contains only zero-sized fields
note: the type is defined here
  --> src/lib.rs:3:1
   |
3  | struct MyPhantom(core::marker::PhantomData<u8>);
   | ^^^^^^^^^^^^^^^^
   = note: `#[warn(improper_ctypes)]` on by default

warning: `playground` (lib) generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.51s
```

This is over-zealous for a few reasons.

1. `PhantomData` on its own is allowed in C structs, so changing this to `_marker: PhantomData<u8>` would solve the warning (but defeats the point of my use case[^1]).

2. `#[repr(transparent)]` is supposed to be treated as the inner type for purposes of ABI, and the `improper_ctypes` lint is about ABI.

3. Using `#[repr(C)]` instead of `#[repr(transparent)]` solves this technically (it shuts the warning up), but seems pretty dodgy, since it's not actually laid out the way C would lay it out anymore (that is, we've guaranteed that `PhantomData` is never going to impact layout at all, and we *haven't* guaranteed this for zero-sized `#[repr(C)]` types which are notably not a real thing in C).

TLDR: this lint fires on `#[repr(transparent)]` wrappers around `PhantomData` when they're used as struct fields. IMO these should be fine since `PhantomData` doesn't emit a warning in that case, and `#[repr(transparent)]` is supposed to defer to the inner type.

[^1]: If you're curious about the real use case I have, a closer but still-basically-minimized version is [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=614935f0bf5a85469cf646f1f89f4ebe). It's important that the type not be a direct `PhantomData` (basically `#[non_exhaustive]` is not really usable for FFI, but C has cases where considers appending new fields to the end of a type to be a non-breaking change).

*(That said, firing this lint on a type passed by pointer is pretty weird anyway, since the indirection means doesn't actually have to be a valid C type (for the ABI), which is the basic complaint behind #66373, which is *not* what I'm complaining about. IMO this shouldn't fire even if `Bar` were being passed by value, as making the field `PhantomData` (instead of a wrapper around one) doesn't fire in that case)*

### PR DESCRIPTION
**Pull request by krtab**
This presents one possible solution to issue: #106629.

This is my first (tentative) contribution to the compiler itself.

I'm looking forward for comments and feedback

Closes: #106629

### CODE REVIEW

File: compiler/rustc_lint/src/types.rs
Comment (davidtwco) [inline compiler/rustc_lint/src/types.rs:891]: Could you attempt to deduplicate this a little with the logic in the `else` block that immediately follows this one?

File: compiler/rustc_lint/src/types.rs
Comment (davidtwco) [inline compiler/rustc_lint/src/types.rs:898]: It's better to follow what happens below where `check_field_type_for_ffi` is invoked and an `FfiPhantom` result is checked for, rather than what you're doing here.

File: compiler/rustc_lint/src/types.rs
Comment (davidtwco) [inline compiler/rustc_lint/src/types.rs:900]: I think we probably should check for this.

**Review by davidtwco — CHANGES_REQUESTED** [review]
This is a good start, thanks for the contribution, left some comments below.

File: compiler/rustc_lint/src/types.rs
Comment (krtab) [reply compiler/rustc_lint/src/types.rs:891]: Good question. This is what I tried originally, but I have always ended up with more akward code that what is here. A notable issue is to output the correct error message. If you delay the emission of `lint_improper_ctypes_struct_zst` until you have gone through the field loop, you end up emitting the wrong error message because of the `r => r` match arm, and trying to remove it led to awkward code iirc.

Comment (davidtwco) [reply compiler/rustc_lint/src/types.rs:891]: It might be okay if another lint was emitted first before `lint_improper_ctypes_struct_zst` - that might not be what we have in the output now, but it would still be correct, I think (because if you fixed that lint, you would then get the `lint_improper_ctypes_struct_zst` after).

**Comment by krtab** [comment]
I think the new version takes into account your points 2 and 3, and I answered 1 above. LMKWYT
@rustbot ready

File: compiler/rustc_lint/src/types.rs
Comment (davidtwco) [reply compiler/rustc_lint/src/types.rs:891]: What about something like this?

```rust
let maybe_default_result = if def.repr().transparent() {
    if let Some(field) = transparent_newtype_field(self.cx.tcx, variant) {
        self.check_field_type_for_ffi(cache, field, substs)
    } else {
        Some(FfiUnsafe { ty, reason: fluent::lint_improper_ctypes_struct_zst, help: None })
    }
} else {
    None
};

let mut all_phantom = !variant.fields.is_empty();
for field in &variant.fields {
    match self.check_field_type_for_ffi(cache, &field, substs) {
        /* what you have */
        r => return maybe_default_result.unwrap_or(r),
    }
}

if all_phantom { FfiPhantom(ty) } else { maybe_default_result.unwrap_or(FfiSafe) }
```

Comment (krtab) [reply compiler/rustc_lint/src/types.rs]: This seems to work fine and to work on my example. Thanks! :)

**Comment by davidtwco** [comment]
There's a test failing, after you fix that then I can approve :)

**Comment by krtab** [comment]
This is a bit above my pay-grade:

The test failing is basically:
```rust
#[repr(transparent)]
enum TransparentEnum<T> {
    Variant(T, std::marker::PhantomData<Z>),
}
```

which fails because it is an enum with a PhantomData, triggering this code path
https://github.com/rust-lang/rust/blob/f5ab431cb7bd2dbf2aa50daa187c8e53ba1419f8/compiler/rustc_lint/src/types.rs#L901-L907

I'm not sure which one is the correct behavior. Should it have been considered not FFI safe since the beginning, or is this PR introducing a new false positive?

**Comment by krtab** [comment]
I've pushed a fix for the failing test but I'll leave it to you to decide which is the correct behavior.

@rustbot ready

File: compiler/rustc_lint/src/types.rs
Comment (davidtwco) [inline compiler/rustc_lint/src/types.rs:885]: I wonder if returning this result is actually what we want - that way we only loop over fields in the "all fields are ZST" case and the "not repr transparent" case, which is what we wanted.

**Comment by krtab** [comment]
It his is ok for you, I guess all is left is for me to squash that into a single commit and mark the PR as not draft.

**Comment by davidtwco** [comment]
> It this is ok for you, I guess all is left is for me to squash that into a single commit and mark the PR as not draft.

Works for me :)

**Comment by krtab** [comment]
Looks like we're good to go!
Thanks a lot for the reviewing work, and cheers to my first contribution to rustc! :partying_face:

**Review by davidtwco — APPROVED** [review]

**Comment by davidtwco** [comment]
In fact, one last thing, can you add a test case for #106629?

**Comment by krtab** [comment]
My bad, I also forgot to do it!
Should be ok now.

**Comment by davidtwco** [comment]
> My bad, I also forgot to do it! Should be ok now.

Tests were moved from `src/test` to `tests/` recently, can you rebase and change it to that? CI will fail otherwise.

**Pull request merged**
